coroutines-guide.md

```diff
@@ -2022,79 +2025,164 @@ The locking in this example is fine-grained, so it pays the price. However, it i
     counter.send(GetCounter(response))
     println("Counter = ${response.await()}")
-    counter.close() // shutdown the actor
+    counter.close()
+    counter.join() // shutdown the actor and wait for it
 }
 ```
 
-> You can get full code [here](core/kotlinx-coroutines-core/test/guide/example-sync-07.kt)
+> You can get full code [here](core/kotlinx-coroutines-core/test/guide/example-actors-01.kt)
 
 <!--- TEST ARBITRARY_TIME
+Preparing to send a lot of inc requests
+CountingActor started
 Completed 1000000 actions in xxx ms
 Counter = 1000000
 -->
 
-It does not matter (for correctness) what context the actor itself is executed in. An actor is
+Note that the same actor, but without `onStart` handler, can be implemented using more compact [actor] builder: 
+```kolin
+fun createCountingActor(): TypedActor<CounterMsg> = actor { message ->
+    when (message) {
+        is IncCounter -> counter++
+        is GetCounter -> message.response.complete(counter)
+    }
+}
+```
+
+ `counter` is not protected by any synchronization, because actor processes messages sequentially one by one.
+It does not matter for correctness what context the actor itself is executed in. An actor is
 a coroutine and a coroutine is executed sequentially, so confinement of the state to the specific coroutine
 works as a solution to the problem of shared mutable state. Indeed, actors may modify their own private state, but can only affect each other through messages (avoiding the need for any locks).
 
 Actor is more efficient than locking under load, because in this case it always has work to do and it does not 
 have to switch to a different context at all.
 
-> Note, that an [actor] coroutine builder is a dual of [produce] coroutine builder. An actor is associated 
-  with the channel that it receives messages from, while a producer is associated with the channel that it 
-  sends elements to.
+### Typesafe actors
+What is usually counterintuitive about Actor-based API is the absence of meaningful method names and requirement to introduce `sealed` classes for every actor
+which processes more than one type of messages. And in any real-world project it's very likely that any actor can process more than only one kind of requests.
+For this purposes we have [Actor] base class. With [Actor], any method with any arguments can act like `send` method of `TypedActor`, the only requirement is to define methods
+using [act][Actor.act].
+
+Let's implement the same `CountingActor` as in the previous example;  
+
+<!--- INCLUDE
+import kotlinx.coroutines.experimental.actors.*
+import kotlinx.coroutines.experimental.guide.sync01.*
+-->
+
+```kotlin
+class CountingActor : Actor() {
+    
+    private var counter: Int = 0
+    
+    override suspend fun onStart() {
+      println("CountingActor started")
+    }
+    
+    suspend fun increment() = act { // <- note act {} extension
+      counter++
+    }
+    
+    suspend fun counter(response: CompletableDeferred<Int>) = act {
```

Comment:
Is parameters as an output the final design?

Should the `counter` method return an `Int`?

https://github.com/Kotlin/kotlinx.coroutines/issues/87#issuecomment-397554104

Reply:
Yes.
We discourage "send and wait" pattern as it's not an actor-ish way to do things, if you need it, you're probably using the wrong abstraction.
Actors come in a system, one actor is no actor. In the real world `counter` method would send its result to another actor (which then updates UI, prints it to the console, whatever).

Nevertheless, for some integrations with existing code it's more or less trivial to write `actAndResponse` on top of `act` in 5-10 lines of code. But we don't want to provide it in the library to avoid misusages